File: sdk-python/agent_phonon/rpc.py
"""双向 JSON-RPC 2.0 peer（asyncio）。每个 device 连接一个。"""
from __future__ import annotations

import asyncio
import json
import uuid
from typing import Any, Awaitable, Callable

RpcHandler = Callable[[str, Any], Awaitable[Any]]


class RpcPeer:
    def __init__(self, send: Callable[[str], Awaitable[None]], handler: RpcHandler) -> None:
        self._send = send
        self._handler = handler
        self._pending: dict[Any, asyncio.Future] = {}
        self._next_id = 1

    async def request(self, method: str, params: Any, timeout: float = 600.0) -> Any:
        rid = self._next_id
        self._next_id += 1
        fut: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending[rid] = fut
        await self._send(json.dumps({"jsonrpc": "2.0", "id": rid, "method": method, "params": params}))
        try:
            return await asyncio.wait_for(fut, timeout)
        finally:
            self._pending.pop(rid, None)

    async def notify(self, method: str, params: Any) -> None:
        await self._send(json.dumps({"jsonrpc": "2.0", "method": method, "params": params}))

    async def handle(self, data: str) -> None:
        try:
            msg = json.loads(data)
        except json.JSONDecodeError:
            return
        # 响应
        if ("result" in msg or "error" in msg) and "id" in msg:
            fut = self._pending.get(msg["id"])
            if fut and not fut.done():
                if "error" in msg and msg["error"]:
                    fut.set_exception(RpcError(msg["error"]))
                else:
                    fut.set_result(msg.get("result"))
            return
        # 请求 / 通知
        method = msg.get("method")
        if isinstance(method, str):
            is_notification = "id" not in msg or msg.get("id") is None
            try:
                result = await self._handler(method, msg.get("params"))
                if not is_notification:
                    await self._send(json.dumps({"jsonrpc": "2.0", "id": msg["id"], "result": result if result is not None else None}))
            except Exception as e:  # noqa: BLE001
                if not is_notification:
                    await self._send(json.dumps({"jsonrpc": "2.0", "id": msg["id"], "error": {"code": -32000, "message": str(e)}}))

    def reject_all(self, reason: str) -> None:
        for fut in self._pending.values():
            if not fut.done():
                fut.set_exception(ConnectionError(reason))
        self._pending.clear()


class RpcError(Exception):
    def __init__(self, error: Any) -> None:
        self.error = error
        msg = error.get("message", str(error)) if isinstance(error, dict) else str(error)
        self.app_code = (error.get("data") or {}).get("appCode") if isinstance(error, dict) else None
        super().__init__(msg)


def new_id() -> str:
    return str(uuid.uuid4())

File: packages/sdk-server-ts/src/rpc.ts
import { randomUUID } from "node:crypto";

/**
 * 双向 JSON-RPC 2.0 peer（server 侧，零 core 依赖）。
 * 每个 device 连接一个。两端皆可作 requester。
 */

export interface Transport {
  send(data: string): void;
  close(): void;
}

export type RpcHandler = (method: string, params: unknown) => Promise<unknown> | unknown;

interface Pending {
  resolve: (v: unknown) => void;
  reject: (e: unknown) => void;
  timer?: ReturnType<typeof setTimeout>;
}

export class RpcPeer {
  private transport: Transport;
  private handler: RpcHandler;
  private pending = new Map<string | number, Pending>();
  private nextId = 1;

  constructor(transport: Transport, handler: RpcHandler) {
    this.transport = transport;
    this.handler = handler;
  }

  request(method: string, params: unknown, timeoutMs = 600000): Promise<unknown> {
    const id = this.nextId++;
    return new Promise((resolve, reject) => {
      const timer = timeoutMs > 0 ? setTimeout(() => { this.pending.delete(id); reject(new Error(`rpc timeout: ${method}`)); }, timeoutMs) : undefined;
      this.pending.set(id, { resolve, reject, timer });
      this.transport.send(JSON.stringify({ jsonrpc: "2.0", id, method, params }));
    });
  }

  notify(method: string, params: unknown): void {
    this.transport.send(JSON.stringify({ jsonrpc: "2.0", method, params }));
  }

  async handle(data: string): Promise<void> {
    let msg: Record<string, unknown>;
    try { msg = JSON.parse(data); } catch { return; }

    // 响应
    if (("result" in msg || "error" in msg) && "id" in msg) {
      const p = this.pending.get(msg.id as string | number);
      if (!p) return;
      this.pending.delete(msg.id as string | number);
      if (p.timer) clearTimeout(p.timer);
      if ("error" in msg && msg.error) p.reject(msg.error);
      else p.resolve((msg as { result: unknown }).result);
      return;
    }

    // 请求 / 通知
    if (typeof msg.method === "string") {
      const isNotification = !("id" in msg) || msg.id === undefined || msg.id === null;
      try {
        const result = await this.handler(msg.method, msg.params);
        if (!isNotification) this.transport.send(JSON.stringify({ jsonrpc: "2.0", id: msg.id, result: result ?? null }));
      } catch (err) {
        if (!isNotification) {
          this.transport.send(JSON.stringify({
            jsonrpc: "2.0", id: msg.id,
            error: { code: -32000, message: (err as Error)?.message ?? "error" },
          }));
        }
      }
    }
  }

  rejectAll(reason: string): void {
    for (const [, p] of this.pending) { if (p.timer) clearTimeout(p.timer); p.reject(new Error(reason)); }
    this.pending.clear();
  }
}

export function newId(): string {
  return randomUUID();
}
